src/FffGcodeWriter.cpp

```diff
@@ -455,6 +455,7 @@ void FffGcodeWriter::processStartingCode(const SliceDataStorage& storage, const
         const RetractionConfig& retraction_config = storage.retraction_config_per_extruder[start_extruder_nr];
         gcode.writeRetraction(retraction_config);
     }
+    gcode.writeExtrusionMode();
```

Comment:
With what extrusion mode is the starting gcode processed?
Should we perhaps set the extrusion mode to absolute extrusion before the processing of the start gcode.

Also I would like to set the extrusion back to normal at the end of a print.

I suggest you change the function to
```
void GCodeExport::writeExtrusionMode(bool relative_extrusion)
```

Reply:
The starting gcode is processed with whatever is the default for the printer.
So I have been adding an M82 to my start code until now to set it to absolute mode as the printer defaults to relative mode.
I'm not sure about outputting an M82 before executing the start code. Alternatively, we could set the extrusion mode before the start code (as well as after) to ensure consistency. i.e. it would look like this:

M82 or M83 depending on extrusion mode
start gcode
M82 or M83 depending on extrusion mode
first layer, etc..

That way, it guarantees that the extrusion mode will be correct during the print and also the user will know that on entering the start gcode, the extrusion mode is what they have specified in the setting.

I don't see how you can set the extrusion mode back to "normal" at the end of the print because what is normal will vary between printers.

Reply:
I think people don't want to change their start gcode because they change the setting Relative Extrusion.

I think we can define the "normal" extrusion mode to be relative for RepRap and absolute for everything else.

Then we can set the extrusion mode to "normal" before the start-gcode and at the end of a print (before the end gcode).

It's important to set the extrusion mode back to normal because not all firmware types set all printing parameters back to what's normal for that machine.

Reply:
> I think people don't want to change their start gcode because they change the setting Relative Extrusion.

That's reasonable.

> I think we can define the "normal" extrusion mode to be relative for RepRap and absolute for everything else.

>Then we can set the extrusion mode to "normal" before the start-gcode and at the end of a print (before the end gcode).

>It's important to set the extrusion mode back to normal because not all firmware types set all printing parameters back to what's normal for that machine.

Hmm, I'm not at all convinced about this.

I can't see how you can define normal in this instance. Whether the printer defaults to relative or absolute extrusion mode is a configurable parameter for some printers.

Reply:
Well the problem was for printers which don't reset relative/absolute extrusion to their default. Those printers will remain in relative extrusion mode if the next print is not sliced with Cura.

If we don't set anything at the end of a print the problem will remain.

If we guess the printer default by the gcode flavor (only relative for RepRap) then we at least solved the problem for those printers where that guess is correct.
It's not a full proof solution, but it least it solves the problem for half of the users who would otherwise experience it.

--------------------------------------------

As for the start gcode:
We could also always set the printer to absolute extrusion before the start gcode so that the start gcode of everyone can remain unchanged.